kore/src/Kore/Step/RewriteStep.hs

```diff
@@ -292,6 +295,21 @@ applyWithFinalizer finalize rules initial = do
     locations = from @_ @SourceLocation . extract
 {-# INLINE applyWithFinalizer #-}
 
+applyRuleWithFinalizer ::
+    forall simplifier.
+    MonadSimplify simplifier =>
+    Finalizer (RulePattern RewritingVariableName) simplifier ->
+    -- | Rewrite rules
+    [RulePattern RewritingVariableName] ->
+    -- | Configuration being rewritten
+    Pattern RewritingVariableName ->
+    simplifier (Results (RulePattern RewritingVariableName))
+applyRuleWithFinalizer finalize rules initial = do
+    results <- applyWithFinalizer finalize rules initial
+    debugRewriteSubstitution initial results
+    return results
+{-# INLINE applyRuleWithFinalizer #-}
```

Comment:
My understanding from the PR description is that you also want to log claim applications, right? I think this should be do-able in the following way:
- don't keep whole `UnifiedRule`s in `DebugRewriteSubstitution`, as I said in the comment above
- add a `From rule UniqueId` constraint [here](https://github.com/kframework/kore/blob/cc94a3f594469011d790d6d89e01069caf13011d/kore/src/Kore/Step/RewriteStep.hs#L171) (like the one for `SourceLocation`)
- add a `From rule UniqueId` constraint in `applyWithFinalizer`
- then I think you'll need to implement `From rule UniqueId` instances for each `rule` the compiler complains about

After these changes you should be able to call `debugRewriteSubstitution` (modified accordingly) in `applyWithFinalizer`, so it will work for both axioms and claims. Please let me know if you run into any problems.